### PR TITLE
perf: use events optimize nacos service discovery

### PR DESCRIPTION
**Pull request by Demogorgon314**
### What this PR does / why we need it:
#4346 

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: apisix/discovery/nacos.lua [hunk 1]
```diff
@@ -80,8 +83,16 @@ local schema = {
...
 
-local _M = {}
+local _M = {
+    version = 0.2,
+}
 
+local function discovery_nacos_callback(data, event, source, pid)
+    applications = data
+    log.notice("update local variable application, event is: ", event,
+            "source: ", source, "server pid:", pid,
```
Comment (spacewander) [inline apisix/discovery/nacos.lua:93]: Need to fix the indent.
Please update other similar places.

File: apisix/discovery/nacos.lua [hunk 2]
```diff
@@ -344,6 +365,18 @@ function _M.init_worker()
         error('invalid nacos configuration: ' .. err)
         return
     end
+
+    events = require("resty.worker.events")
+    events_list = events.event_list(
+            "discovery_nacos_update_application",
+            "updating"
```
Comment (spacewander) [inline apisix/discovery/nacos.lua:372]: Need to add `workers(4);` to https://github.com/apache/apisix/blob/0d7fa659d557a83224879cd1bfa823deff606bf7/t/discovery/nacos.t#L23 to check if it works

Don't forget to update https://github.com/apache/apisix/blob/master/docs/en/latest/discovery/nacos.md#service-discovery-via-nacos

**Comment by tzssangglass** [comment]
I think additional test cases are needed

File: apisix/discovery/nacos.lua [hunk 3]
```diff
@@ -311,7 +322,17 @@ local function fetch_full_registry(premature)
             })
         end
     end
+    local new_apps_md5sum = ngx.md5(core.json.encode(up_apps))
+    local old_apps_md5sum = ngx.md5(core.json.encode(applications));
```
Comment (tokers) [inline apisix/discovery/nacos.lua:326]:
```suggestion
    local old_apps_md5sum = ngx.md5(core.json.encode(applications))
```

**Review by tokers — APPROVED** [review]

File: apisix/discovery/nacos.lua [hunk 4]
```diff
@@ -311,7 +322,17 @@ local function fetch_full_registry(premature)
...
+    if new_apps_md5sum == old_apps_md5sum then
+        return
+    end
     applications = up_apps
+    local ok, err = events.post(events_list._source, events_list.updating, applications)
+    if not ok then
+        log.error("post_event failure with ", events_list._source,
+                ", update application error: ", err)
```
Comment (tokers) [inline apisix/discovery/nacos.lua:334]: Bad indent.

**Review by tokers — CHANGES_REQUESTED** [review]

**Review by membphis — COMMENTED** [review]
missing test cases

File: apisix/discovery/nacos.lua (hunk 4 above)
Comment (Demogorgon314) [reply apisix/discovery/nacos.lua]: fixed.

File: docs/en/latest/discovery/nacos.md [hunk 5]
```diff
@@ -27,7 +27,7 @@ This is experimental discovery module for Nacos.
 
 The performance of this module needs to be improved:
 
-1. avoid synchroning configuration in each workers. You can refer the implementation in `consul_kv.lua`.
+1. ~~avoid synchroning configuration in each workers. You can refer the implementation in `consul_kv.lua`.~~
```
Comment (spacewander) [inline docs/en/latest/discovery/nacos.md:30]: We can directly remove it.

File: apisix/discovery/nacos.lua [hunk 6]
```diff
@@ -80,8 +83,16 @@ local schema = {
 }
 
 
-local _M = {}
+local _M = {
+    version = 0.2,
```
Comment (spacewander) [inline apisix/discovery/nacos.lua:87]: We don't need to add the version number.

Comment (Demogorgon314) [reply apisix/discovery/nacos.lua]: Thx. I fixed.

**Review by spacewander — APPROVED** [review]

**Review by tokers — APPROVED** [review]

**Pull request merged**